stdlib/uchar.ml

```diff
@@ -20,41 +20,46 @@ let err_no_succ = "U+10FFFF has no successor"
 let err_not_sv i = format_int "%X" i ^ " is not a Unicode scalar value"
 let err_not_latin1 u = "U+" ^ format_int "%04X" u ^ " is not a latin1 character"
 
-type t = int
+type t = uchar
+external unsafe_of_int : int -> t = "%identity"
+external to_int : t -> int = "%identity"
 
-let min = 0x0000
-let max = 0x10FFFF
-let lo_bound = 0xD7FF
-let hi_bound = 0xE000
+let min = unsafe_of_int 0x0000
+let max = unsafe_of_int 0x10FFFF
+let lo_bound = unsafe_of_int 0xD7FF
+let hi_bound = unsafe_of_int 0xE000
 
-let bom = 0xFEFF
-let rep = 0xFFFD
+let bom = unsafe_of_int 0xFEFF
+let rep = unsafe_of_int 0xFFFD
 
 let succ u =
   if u = lo_bound then hi_bound else
   if u = max then invalid_arg err_no_succ else
-  u + 1
+  unsafe_of_int (to_int u + 1)
 
 let pred u =
   if u = hi_bound then lo_bound else
   if u = min then invalid_arg err_no_pred else
-  u - 1
+  unsafe_of_int (to_int u - 1)
 
-let is_valid i = (min <= i && i <= lo_bound) || (hi_bound <= i && i <= max)
-let of_int i = if is_valid i then i else invalid_arg (err_not_sv i)
-external unsafe_of_int : int -> t = "%identity"
-external to_int : t -> int = "%identity"
+let is_valid i =
+  let i = unsafe_of_int i in
+  (min <= i && i <= lo_bound) || (hi_bound <= i && i <= max)
```

Comment:
This seems a bit dicey: the compiler knows (at least in some places, like the switcher stuff in pattern compilation), that uchar type is between Uchar.min and Uchar.max. So the compiler would be within its right to observe that `min <= i`  and `i <= max` are redundant, and remove them.

```suggestion
  (to_int min <= i && i <= to_int lo_bound) || (to_int hi_bound <= i && i <= to_int max)
```